Makefile

gucio321 opened the pull request:
Hi!
I wrote a simple makefile to get rid of just.
feel free to suggest changes as I'm not really familiar with make :smile:

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+NAME=cimgui-go Code Generator
+GOCMD=LC_ALL=C go
+TIMEOUT=5
```
Why do you need TIMEOUT?

gucio321 replied:
I don't know to be honest.
It was in the Makefile I was basing on - will remove.

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+NAME=cimgui-go Code Generator
+GOCMD=LC_ALL=C go
```
Why do you need this?

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+# go tools
+export PATH := ./bin:$(PATH)
+export GO111MODULE := on
+export GOPROXY = https://proxy.golang.org,direct
```
Not needed

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+
+# go tools
+export PATH := ./bin:$(PATH)
+export GO111MODULE := on
```
Not needed

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+TIMEOUT=5
+
+# go tools
+export PATH := ./bin:$(PATH)
```
Why do you need to change PATH?

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+export GO111MODULE := on
+export GOPROXY = https://proxy.golang.org,direct
+
+.PHONY: all setup cimgui cimplot generate
```
Let's do it like this:

```make
.PHONY: all
all:
        ...
```

for all targets. Putting `PHONY` in a separate place is confusing.

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+
+## setup: pulls down dependencies; builds codegen programm
+setup:
+	$(GOCMD) build -o ./cmd/codegen/build/codegen ./cmd/codegen
```
Do you really need a "setup" step? Can't it be just "go run"?

gucio321 replied:
idk, it was this way in justfile so I didn't change anything so far.

gucio321 replied:
I think, that `go run` is a good idea, but i'd leave the step and put here `go get -d ./...` for now
an maybe in th efuture, if we'll need to use `go generate` or something like that, we may put it there.

eliasdaler reviewed (changes requested):
The whole thing with "build directory" is not how things are done in Make usually. It also makes paths really confusing. Everything should be done in the main repo's directory and all paths should be specified relative to it. It will be MUCH easier to follow what's going on.

gucio321 commented:
@eliasdaler thank you for your review!
I'll try to fix all the issues now ;-)

gucio321 commented:
> The whole thing with "build directory" is not how things are done in Make usually. It also makes paths really confusing. Everything should be done in the main repo's directory and all paths should be specified relative to it. It will be MUCH easier to follow what's going on.

ok @eliasdaler, I've changed the code a bit and I suppose it fits your suggestion now. Let me know what do you think about the current implementation, especially about the `setup` rule I've changed a bit.

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,47 @@
+	go get -v -d ./...
+
+## _gencode: is an internal rule. It generates binding basing on env variables set in imgui and implot rules
+_generate: setup
```
Looks like this needs to be a function [see example here](https://coderwall.com/p/cezf6g/define-your-own-function-in-a-makefile), then you'll be able to write cimplot/cimgui targets easier: they'll just set set needed variables and then call the function

gucio321 replied:
Oh, I can't imagine how did i miss it while looking around makefile's documentation...
Thank you very much!

gucio321 replied:
> and maybe in th efuture, if we'll need to use go generate or something like that, we may put it there.

also, I'm wondering if we shouldn't obtain `gofmt` here? User isn't obligated to have it installed tbh.

eliasdaler commented:
Much better, I have only one more request - change `_generate` from target to function, please.

eliasdaler commented on Makefile:
```diff
@@ -0,0 +1,45 @@
+
+.PHONY: setup
+## setup: pulls down dependencies.
+setup:
```
Do you actually need this? "go run" pulls dependencies automatically.

gucio321 replied:
> and maybe in th efuture, if we'll need to use go generate or something like that, we may put it there.

also, I'm wondering if we shouldn't obtain gofmt here? User isn't obligated to have it installed tbh.

eliasdaler replied:
> and maybe in the future, if we'll need to use go generate, stringer or something like that, we may put it there.
I don't like leaving unnecessary things "for the future" - we can easily add the targets we need later. It will be cleaner.

> also, I'm wondering if we shouldn't obtain gofmt here?
Go ships with gofmt, so this is not needed.

gucio321 replied:
okey, sounds fair

eliasdaler reviewed (changes requested):
Sorry, noticed one more thing

gucio321 commented:
@eliasdaler what do you think about the lates version? is it ok now?

eliasdaler commented:
Sorry, forgot to merge :)
